File: Server/src/main/java/com/lfj/messenger/server/dao/ChatMemberDAO.java
package com.lfj.messenger.server.dao;

import com.lfj.messenger.dto.datatype.server.ChatMemberDTO;
import com.lfj.messenger.dto.datatype.server.UserDTO;
import com.lfj.messenger.dto.types.Role;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import javax.sql.DataSource;
import java.sql.*;
import java.time.Instant;
import java.util.*;
import java.util.function.Function;
import java.util.stream.Collectors;

public class ChatMemberDAO {
    private DataSource dataSource;
    private Logger logger;

    private final static String INSERT = "INSERT INTO chat_members_table (chat_id, user_id, role, joined_at) VALUES (?, ?, ?, ?)";
    private final static String SELECT_BY_CHAT_ID = "SELECT user_id, role, joined_at FROM chat_members_table WHERE chat_id = ?";
    private final static String SELECT_USER_ID_BY_CHAT_ID = "SELECT user_id FROM chat_members_table WHERE chat_id = ?";

    private ChatMemberDAO(){  }
    public ChatMemberDAO(DataSource dataSource){
        this.logger = LoggerFactory.getLogger(ChatMemberDAO.class);
        this.dataSource = dataSource;
    }

    public void addMemberForPrivateChat(UUID chatId, Set<UUID> members){
        try(Connection connection = this.dataSource.getConnection(); PreparedStatement stmt = connection.prepareStatement(INSERT)){
            for(UUID member : members){
                Instant instant = Instant.now();
                stmt.setObject(1, chatId);
                stmt.setObject(2, member);
                stmt.setString(3, Role.MEMBER.name());
                stmt.setTimestamp(4, Timestamp.from(instant));
                stmt.executeUpdate();
            }
        }catch (SQLException e){
            logger.error("Database error", e);
        }
    }

    public Optional<ChatMemberDTO> selectChatMembersById(UUID chatId, Set<UserDTO> users){
        Map<Role, Set<ChatMemberDTO.Member>> members = new HashMap<>();
        Map<UUID, UserDTO> map = users.stream().collect(Collectors.toMap(UserDTO::userId, Function.identity()));
        try(Connection connection = this.dataSource.getConnection(); PreparedStatement stmt = connection.prepareStatement(SELECT_BY_CHAT_ID)) {
            stmt.setObject(1, chatId);
            try(ResultSet rs = stmt.executeQuery()){
                while(rs.next()){
                    Role role = Role.valueOf(rs.getString("role"));
                    UUID userId = rs.getObject("user_id", UUID.class);
                    UserDTO user = map.get(userId);
                    Instant instant = rs.getTimestamp("joined_at").toInstant();
                    members.computeIfAbsent(role, v -> new HashSet<>()).add(new ChatMemberDTO.Member(user, instant));
                }
            }
            return Optional.of(new ChatMemberDTO(chatId, members));
        }catch (SQLException e){
            logger.error("Database error", e);
        }
        return Optional.empty();
    }
    public Optional<Set<UUID>> getUsersId(UUID chatId){
        try(Connection connection = this.dataSource.getConnection(); PreparedStatement stmt = connection.prepareStatement(SELECT_USER_ID_BY_CHAT_ID)){
            Set<UUID> userIds = new HashSet<>();
            stmt.setObject(1, chatId);
            try(ResultSet rs = stmt.executeQuery()) {
                while (rs.next()){
                    userIds.add(rs.getObject("user_id", UUID.class));
                }
            }
            return Optional.of(userIds);
        } catch (SQLException e) {
            logger.error("Database error", e);
        }
        return Optional.empty();
    }
}

File: Server/src/main/kotlin/com/lfj/messfox/server/dao/ChatMemberDao.kt
package com.lfj.messfox.server.dao

import com.lfj.messfox.protocol.datatype.ChatMember
import com.lfj.messfox.protocol.type.Role
import org.slf4j.Logger
import org.slf4j.LoggerFactory
import java.sql.SQLException
import java.sql.Timestamp
import java.time.Instant
import java.util.*
import javax.sql.DataSource
import kotlin.reflect.KClass

interface ChatMemberDataTable {
    fun addMemberForChat(chatId: UUID, userId: UUID, role: Role) : Result<ChatMember>
    fun addMembersForChat(chatId: UUID, userIds: List<UUID>, role: Role) : List<ChatMember>
    fun deleteMemberForChat(chatId: UUID, userId: UUID)
    fun selectUserByChatId(chatId: UUID) : List<ChatMember>
    fun selectChatsIdForUserId(userId: UUID) : List<UUID>

    interface CompanionsSQLRequest{
        val ADD_MEMBER_FOR_CHAT: String
        val REMOVE_MEMBER_FOR_CHAT: String
        val SELECT_MEMBERS_FOR_CHAT: String
        val SELECT_CHATS_FOR_USER: String
    }
    open class AbstractChatMembersDao <T : ChatMemberDataTable> (open val dataSource: DataSource, val sqlRequest: CompanionsSQLRequest, clazz: KClass<T>) : ChatMemberDataTable{
        private val logger: Logger = LoggerFactory.getLogger(clazz.java)
        override fun addMemberForChat(chatId: UUID, userId: UUID, role: Role) : Result<ChatMember> {
            return try{
                dataSource.connection.use{ conn ->
                    conn.prepareStatement(sqlRequest.ADD_MEMBER_FOR_CHAT).use { stmt ->
                        val instant: Instant = Instant.now()
                        stmt.setObject(1, chatId)
                        stmt.setObject(2, userId)
                        stmt.setString(3, role.name)
                        stmt.setTimestamp(4, Timestamp.from(instant))
                        stmt.executeUpdate()
                        Result.success(ChatMember(userId, role, instant))
                    }
                }
            }catch (e: SQLException){
                logger.error("Error ChatMembersDao", e)
                Result.failure(e)
            }
        }

        override fun addMembersForChat(chatId: UUID, userIds: List<UUID>, role: Role): List<ChatMember> {
            val mutableList: MutableList<ChatMember> = mutableListOf()
            return try{
                dataSource.connection.use{ conn ->
                    conn.prepareStatement(sqlRequest.ADD_MEMBER_FOR_CHAT).use{ stmt ->
                        userIds.forEach {
                            val instant: Instant = Instant.now()
                            stmt.setObject(1, chatId)
                            stmt.setObject(2, it)
                            stmt.setString(3, role.name)
                            stmt.setTimestamp(4, Timestamp.from(instant))
                            stmt.executeUpdate()
                            mutableList.add(ChatMember(it, role, instant))
                        }
                        mutableList.toList()
                    }
                }
            }catch (e: SQLException){
                logger.error("Error ChatMembersDao", e)
                listOf()
            }
        }
        override fun deleteMemberForChat(chatId: UUID, userId: UUID) {
            try {
                dataSource.connection.use { conn ->
                    conn.prepareStatement(sqlRequest.REMOVE_MEMBER_FOR_CHAT).use { stmt ->
                        stmt.setObject(1, chatId)
                        stmt.setObject(2, userId)
                        stmt.executeUpdate()
                    }
                }
            }catch (e: SQLException){
                logger.error("Error ChatMemberDao", e)
            }
        }
        override fun selectUserByChatId(chatId: UUID) : List<ChatMember> {
            val mutableList: MutableList<ChatMember> = mutableListOf()
            try { dataSource.connection.use { conn ->
                    conn.prepareStatement(sqlRequest.SELECT_MEMBERS_FOR_CHAT).use{ stmt ->
                        stmt.setObject(1, chatId)
                        stmt.executeQuery().use{ rs ->
                            while(rs.next()) mutableList.add(ChatMember(rs.getObject("user_id", UUID::class.java), Role.valueOf(rs.getString("role")), rs.getTimestamp("joined_at").toInstant()))
                        }
                        return mutableList.toList()
                    }
                }
            }catch (e: SQLException){
                logger.error("Error ChatMemberDao", e)
                return listOf()
            }
        }
        override fun selectChatsIdForUserId(userId: UUID) : List<UUID> {
            val mutableList: MutableList<UUID> = mutableListOf()
            try{ dataSource.connection.use{ conn ->
                    conn.prepareStatement(sqlRequest.SELECT_CHATS_FOR_USER).use{ stmt ->
                        stmt.setObject(1, userId)
                        stmt.executeQuery().use{ rs ->
                            while(rs.next()) mutableList.add(rs.getObject("chatId", UUID::class.java))
                        }
                        return mutableList
                    }
                }
            }catch (e: SQLException){
                logger.error("Error ChatMemberDao", e)
                return listOf()
            }
        }
    }
    class H2ChatMemberDao(override val dataSource: DataSource) : AbstractChatMembersDao<H2ChatMemberDao>(dataSource, h2SQLRequest, H2ChatMemberDao::class) {
        private companion object h2SQLRequest : CompanionsSQLRequest {
            override val ADD_MEMBER_FOR_CHAT = """
                INSERT INTO chat_members_table(chat_id, user_id, role, joined_at) 
                VALUES (?, ?, ?, ?)
            """.trimIndent()
            override val REMOVE_MEMBER_FOR_CHAT = """
                DELETE FROM chat_members_table
                WHERE chat_id = ? AND user_id = ?
            """.trimIndent()
            override val SELECT_MEMBERS_FOR_CHAT = """
                SELECT user_id FROM chat_members_table
                WHERE chat_id = ?
            """.trimIndent()
            override val SELECT_CHATS_FOR_USER = """
                SELECT chat_id FROM chat_members_table
                WHERE user_id = ?
            """.trimIndent()
        }
    }
}